src/handlers.rs

```diff
@@ -440,56 +442,85 @@ fn handle_make_temp_dir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
+    Ok(Ready(Ok(v))) => Ok(v.into()),
+    Ok(Ready(Err(err))) => Err(err),
+    Ok(NotReady) => Ok(NotReady),
+    Err(_) => panic!("blocking error"),
+  }
+}
+
+// TODO Do not use macro for the blocking function.. We should instead be able
+// to do this with a normal function, but there seems to some type system
+// issues. The type of this function should be something like this:
+//   fn blocking<F>(is_sync: bool, f: F) -> Box<Op>
+//   where F: FnOnce() -> DenoResult<Buf>
+macro_rules! blocking {
+  ($is_sync:expr,$fn:expr) => {
+    if $is_sync {
+      // If synchronous, execute the function immediately on the main thread.
+      Box::new(futures::future::result($fn()))
+    } else {
+      // Otherwise dispatch to thread pool.
+      Box::new(poll_fn(move || convert_blocking($fn)))
+    }
+  };
+}
+
 fn handle_mkdir(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
   let msg = base.msg_as_mkdir().unwrap();
   let mode = msg.mode();
-  let path = msg.path().unwrap();
+  let path = String::from(msg.path().unwrap());
 
   if !state.flags.allow_write {
     return odd_future(permission_denied());
   }
-  // TODO Use tokio_threadpool.
-  Box::new(futures::future::result(|| -> OpResult {
+
+  blocking!(base.sync(), || {
     debug!("handle_mkdir {}", path);
-    deno_fs::mkdir(Path::new(path), mode)?;
+    deno_fs::mkdir(Path::new(&path), mode)?;
     Ok(empty_buf())
-  }()))
+  })
 }
 
 fn handle_remove(state: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
   let msg = base.msg_as_remove().unwrap();
-  let path = msg.path().unwrap();
+  let path = PathBuf::from(msg.path().unwrap());
   let recursive = msg.recursive();
-
   if !state.flags.allow_write {
     return odd_future(permission_denied());
   }
-  // TODO Use tokio_threadpool.
-  Box::new(futures::future::result(|| -> OpResult {
-    debug!("handle_remove {}", path);
-    let path_ = Path::new(&path);
-    let metadata = fs::metadata(&path_)?;
+  blocking!(base.sync(), || {
+    debug!("handle_remove {}", path.display());
+    let metadata = fs::metadata(&path)?;
     if metadata.is_file() {
-      fs::remove_file(&path_)?;
+      fs::remove_file(&path)?;
     } else {
       if recursive {
-        remove_dir_all(&path_)?;
+        remove_dir_all(&path)?;
       } else {
-        fs::remove_dir(&path_)?;
+        fs::remove_dir(&path)?;
       }
     }
     Ok(empty_buf())
-  }()))
+  })
 }
 
 // Prototype https://github.com/denoland/isolate/blob/golang/os.go#L171-L184
 fn handle_read_file(_config: Arc<IsolateState>, base: &msg::Base) -> Box<Op> {
   let msg = base.msg_as_read_file().unwrap();
   let cmd_id = base.cmd_id();
-  let filename = String::from(msg.filename().unwrap());
-  Box::new(futures::future::result(|| -> OpResult {
-    debug!("handle_read_file {}", filename);
-    let vec = fs::read(Path::new(&filename))?;
+  let filename = PathBuf::from(msg.filename().unwrap());
+  debug!("handle_read_file {}", filename.display());
+  blocking!(base.sync(), || {
```

Comment:
I like that it is this easy to implement both sync and async versions of the same function.